OneCommunityGlobal/HighestGoodNetworkApp
# yihan create editable four-letter team code option

# Description
<img width="746" alt="image" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94303659/8e82891d-45c0-4a68-9ee1-7737697b2fc2">

## Related PRS (if any):
This frontend PR is related to the [#480](https://github.com/OneCommunityGlobal/HGNRest/pull/480) backend PR.

## Main changes explained:
- Update file TeamsTab.jsx and UserProfile.jsx for adding Team Code section under teams tab.
- Update file WeeklySummariesReport.jsx and FormattedReport.jsx for adding Team Code Input field and code and color filters on weekly summaries report page. 
- Create file BioFunction.jsx and TotalValidSummaries.jsx for spliting bioFunction and getTotalValidWeeklySummaries (which cause `rendered fewer hooks than expected` error ) from FormattedReport. 

## How to test:
1. check into current branch on both front and back end.
2. do `npm run start:local` to run this PR locally
3. log as Admin/Owner user
4. test "Need the code to be editable from the Weekly Summary Reports page"
a. go to Reports → Weekly Summaries Report → edit team code field (on the far right of Name) → click outside the input field
b. refresh page or go to user's profile page to see if the code is saved successfully.
6. test "Need the code to be editable from a person’s Profile Page"
a. go to any user's profile page → Teams Tab → edit Team Code section → click on "save changes" button
b. verify the team code is saved successfully.
8. test "Need the page to be easily filterable by multiple codes and/or colors" and "See how many people are already assigned to each existing code AND color"
a. go to Reports → Weekly Summaries Report → filter weekly summaries by two filters.
b. verify the page changes and displays correct summaries. Besides, see if the number of "Total Team Members" is correct.

## Screenshots or videos of changes:
1. Edit team code on user profile page and weekly summaries page.
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94303659/ce11719c-430a-43bd-bbae-5841aa1e565a

2. Test team code and color filters.
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/94303659/028b6133-cc54-4bbc-9234-916a8d230538

## Note:
If you edit Team code/Bio announcement/Total Valid Weekly Summaries on weekly summaries page, then use filters, you'll see the corresponding data doesn't update. But it acctually has been saved successfully, and will update if you refresh the page. At first, I was thinking load the weekly summaries from database everytime we use the filter. But after did some tests, I found that load summaries repeatedly slows the system performance and brings really bad user experience.
